ipaserver/install/server/upgrade.py

```diff
@@ -1485,14 +1485,20 @@ def add_default_caacl(ca):
 def setup_pkinit(krb):
     root_logger.info("[Setup PKINIT]")
 
-    if not api.Command.ca_is_enabled()['result']:
-        root_logger.info("CA is not enabled")
-        return
+    pkinit_state = krbinstance.get_local_pkinit_state()
+
+    if pkinit_state not in (krbinstance.PKINIT_FULL,
+                            krbinstance.PKINIT_EXTERNAL):
 
-    if not os.path.exists(paths.KDC_CERT):
-        root_logger.info("Requesting PKINIT certificate")
-        krb.setup_pkinit()
+        pkinit_setup_handler = krb.setup_local_pkinit
 
+        if api.Command.ca_is_enabled()['result']:
+            pkinit_setup_handler = krb.setup_full_pkinit
```

Comment:
When we designed this, the idea was that on --no-pkinit we would create an "external" self-signed certificate. This would be like local, but would prevent upgrades because it is not tracked by a CA certmonger can request certs from.

If "external" can't be used, then you have to track somewhere that the admin explicitly passed the --no-pkinit flag and prevent upgrades.

In general we *do* want autometic upgrades unless the admin explicitly and unequivocably set a --no-pkinit flag.

The other option is to completely ignore --no-pkinit and mark those flags as deprecated and document that, if an dmin doesn't want pkinit at all they will have to replace the pkinit certs with certs of their own that work only locally.